File: leetcode/684. Redundant Connection.cc
// Copyright 2017 Qi Wang
// Date: 2017-10-13
// Refer to: https://discuss.leetcode.com/topic/104729/10-line-java-solution-union-find
class Solution {
 public:
  vector<int> findRedundantConnection(vector<vector<int>>& edges) {
    vector<int> parent(edges.size() + 1);
    for (int i = 1; i < parent.size(); ++i) {
      parent[i] = i;
    }
    for (const auto& edge : edges) {
      int i = UnionFind(&parent, edge[0]);
      int j = UnionFind(&parent, edge[1]);
      if (i == j) {
        return {edge[0], edge[1]};
      } else {
        parent[j] = i;
      }
    }
    return {0, 0};  // comfort the compiler
  }
 
 private:
  int UnionFind(vector<int>* parent, int x) const {
    return (*parent)[x] == x ? x
                             : (*parent)[x] = UnionFind(parent, (*parent)[x]);
  }
};

File: leetcode/684. Redundant Connection.java
// Date: 2020-09-05
class Solution {
  public int[] findRedundantConnection(int[][] edges) {
    int n = edges.length;
    int[] p = new int[n + 1];
    for (int i = 1; i <= n; ++i) {
      p[i] = i;
    }
    for (int[] edge : edges) {
      int u = edge[0];
      int v = edge[1];
      int x = parent(p, u);
      int y = parent(p, v);
      if (x == y) {
        return new int[] {u, v};
      } else {
        p[y] = x;
      }
    }
    return new int[]{0, 0};
  }

  private int parent(int[] p, int n) {
    return p[n] == n ? n : (p[n] = parent(p, p[n]));
  }
}
